[article] Resource Optimization Guide (#1342)
qdrant/landing_page: qdrant-landing/content/articles/vector-search-resource-optimization.md

|----------------------|------------------------------------------------------------------------------------------------------------------|
| `type`             | Specifies the quantized vector type (currently supports only int8).                                              |
| `quantile`         | Sets bounds for quantization, excluding outliers. For example, 0.99 excludes the top 1% of extreme values to maintain better accuracy. |
| `always_ram `      | Keeps quantized vectors in RAM to speed up searches.                                                             |

Adjust these settings to strike the right balance between precision and efficiency for your specific workload.

---
Learn More about [**Scalar Quantization**](/documentation/guides/quantization/)

---

#### Binary Quantization

**Binary quantization** takes scalar quantization to the next level by compressing each vector component into just **a single bit**. This method achieves unparalleled memory efficiency and query speed, reducing memory usage by a factor of 32 and enabling searches up to 40x faster.

#### **Benefits of Binary Quantization:**

> Binary quantization is ideal for high-dimensional datasets and compatible embedding models, where compression and speed are paramount. 

Reviewer: ```suggestion
Binary quantization is ideal for large-scale datasets and compatible embedding models, where compression and speed are paramount.
```

The datasets are not highly-dimensional themselves, but BQ works better for high-dimensional vectors. However, it should work best when the embedding dimensions are centered around zero, but that's not that important here.

**Figure 5:** This method causes maximum compression. It reduces memory usage by 32x and speeds up searches by up to 40x.

<img src="/articles_data/vector-search-resource-optimization/binary-quantization.png" alt="binary-quantization" style="width: 75%;">

| Benefit                          | Description                                                                                                      |
|----------------------------------|------------------------------------------------------------------------------------------------------------------|
| **Efficient similarity calculations** | Emulates Hamming distance through dot product comparisons, making it fast and effective. |

> | **Perfect for high-dimensional vectors** | Works well with embedding models like OpenAI’s text-embedding-ada-002 or Cohere’s embed-english-v2.0. |

Reviewer: ```suggestion
| **Perfect for high-dimensional vectors** | Works well with embedding models like OpenAI’s text-embedding-ada-002 or Cohere’s embed-english-v3.0. |
```

There is already a more recent model from Cohere, so it's better to mention it instead. It's also fine to use with BQ.

Efficiently managing large datasets in distributed systems like Qdrant requires smart strategies for data isolation. **Multitenancy** and **Sharding** are essential tools to help you handle high volumes of user-specific data while maintaining performance and scalability.

#### Multitenancy

**Multitenancy** is a software architecture where multiple independent users (or tenants) share the same resources or environment. In Qdrant, a single collection with logical partitioning is often the most efficient setup for multitenant use cases.

**Figure 5:** Each individual vector is assigned a specific payload that denotes which tenant it belongs to. This is how a large number of different tenants can share a single Qdrant collection.

<img src="/articles_data/vector-search-resource-optimization/multitenancy.png" alt="multitenancy" style="width: 75%;">

**Why Choose Multitenancy?**

- **Logical Isolation**: Ensures each tenant’s data remains separate while residing in the same collection.
- **Minimized Overhead**: Reduces resource consumption compared to maintaining separate collections for each user.
- **Scalability**: Handles high user volumes without compromising performance.

> Here’s how you can implement multitenancy efficiently in Qdrant:
> 
> ```python
> client.upsert(
>    collection_name="{collection_name}",
>    points=[
>        models.PointStruct(
>            id=1,
>            payload={"group_id": "user_1"},
>            vector=[0.9, 0.1, 0.1],
>        ),
> 
>        models.PointStruct(
>            id=2,
>            payload={"group_id": "user_2"},
>            vector=[0.5, 0.9, 0.4],
>        )
>    ]
> )
> ```

Reviewer: ``````suggestion
Here’s how you can implement multitenancy efficiently in Qdrant:

```python
client.create_payload_index(
    collection_name="{collection_name}",
    field_name="group_id",
    field_schema=models.KeywordIndexParams(
        type="keyword",
        is_tenant=True,
    ),
)
```

Creating a keyword payload index, with the `is_tenant` parameter set to `True`, modifies the way the vectors will be logically stored. Storage structure will be organized to co-locate vectors of the same tenant together.

Now, each point stored in Qdrant should have the `group_id` payload attribute set:

```python
client.upsert(
   collection_name="{collection_name}",
   points=[
       models.PointStruct(
           id=1,
           payload={"group_id": "user_1"},
           vector=[0.9, 0.1, 0.1],
       ),

       models.PointStruct(
           id=2,
           payload={"group_id": "user_2"},
           vector=[0.5, 0.9, 0.4],
       )
   ]
)
```
``````

<img src="/articles_data/vector-search-resource-optimization/user-defined-sharding.png" alt="user-defined-sharding" style="width: 75%;">

**Example:**

```python
client.create_collection(
    collection_name="my_custom_sharded_collection",
    shard_number=1,
    sharding_method=models.ShardingMethod.CUSTOM
)
client.create_shard_key("my_custom_sharded_collection", "tenant_id")
```

---

When implementing user-defined sharding in Qdrant, two key parameters are critical to achieving efficient data distribution:

1. **Shard Key**:
    
    The shard key determines how data points are distributed across shards. For example, using a key like `tenant_id` allows you to control how Qdrant partitions the data. Each data point added to the collection will be assigned to a shard based on the value of this key, ensuring logical isolation of data.
    
2. **Shard Number**:
    
    This defines the total number of physical shards for each shard key, influencing resource allocation and query performance.
    
Here’s how you can add a data point to a collection with user-defined sharding:

> ```python
> client.upsert( collection_name="my_collection", 
> points=[models.PointStruct(id=1111, vector=[0.1, 0.2, 0.3])], 
> shard_key_selector="tenant_1" )
> ```

Reviewer: ``````suggestion
```python
client.upsert(
    collection_name="my_custom_sharded_collection", 
    points=[
        models.PointStruct(
            id=1111, 
            vector=[0.1, 0.2, 0.3]
        )
    ], 
    shard_key_selector="tenant_1"
)
```
``````

Adjusted the collection name and code formatting.

---

> This code assigns the point to a specific shard based on the tenant_1 shard key, ensuring proper data placement.

Reviewer: ```suggestion
This code assigns the point to a specific shard based on the `tenant_1` shard key, ensuring proper data placement.
```

| **Avoid Re-Sharding**           | Re-sharding can be complex and requires creating a new collection, so it's best to plan ahead to minimize disruptions. |

Learn more about [**Sharding in Distributed Deployment**](/documentation/guides/distributed_deployment/)

---

## Query Optimization
![qdrant](/articles_data/vector-search-resource-optimization/query.png)
Improving vector database performance is critical when dealing with large datasets and complex queries. By leveraging techniques like **filtering**, **batch processing**, **reranking**, **rescoring**, and **oversampling**, so you can ensure fast response times and maintain efficiency even at scale.

#### Filtering

Filtering allows you to select only the required fields in your query results. By limiting the output size, you can significantly reduce response time and improve performance.

The filterable vector index is Qdrant's solves pre and post-filtering problems by adding specialized links to the search graph. It aims to maintain the speed advantages of vector search while allowing for precise filtering, addressing the inefficiencies that can occur when applying filters after the vector search.

**Example:**

> ```python
> results = client.search( collection_name="my_collection", 
> query_vector=[0.1, 0.2, 0.3], limit=10, 
> with_payload={"include": ["category"]})
> ```

Reviewer: ``````suggestion
```python
results = client.search(
    collection_name="my_collection", 
    query_vector=[0.1, 0.2, 0.3], 
    limit=10, 
    with_payload={"include": ["category"]}
)
```
``````

Adjust formatting.

            oversampling=2  # Retrieves extra candidates for rescoring
        )
    ),
    limit=4  # Desired number of final results
)
```
---
Learn more about [**Rescoring**](/articles/what-is-vector-quantization/#3-rescoring-with-original-vectors).

---

#### Reranking

Reranking adjusts the order of search results based on additional criteria, ensuring the most relevant results are prioritized. 

This method is about taking the results from different search methods and reordering them based on some additional processing using the content of the documents, not just the scores. This processing may rely on an additional neural model, such as a cross-encoder which would be inefficient enough to be used on the whole dataset. 

<img src="/articles_data/vector-search-resource-optimization/reranking.png" alt="reranking" style="width: 75%;">

These methods are practically applicable only when used on a smaller subset of candidates returned by the faster search methods. Late interaction models, such as ColBERT, are way more efficient in this case, as they can be used to rerank the candidates without the need to access all the documents in the collection.

**Example:**

> ```python
> client.query_points(
>         "collection-name",
>        prefetch=prefetch, # Previous results
>        query=late_vectors, # Colbert converted query
>        using="colbertv2.0",
>        with_payload=True,
>        limit=10,
> )
> ```

Reviewer: ```suggestion
```python
client.query_points(
       "collection-name",
       prefetch=prefetch, # Previous results
       query=late_vectors, # Colbert converted query
       using="colbertv2.0",
       with_payload=True,
       limit=10,
)
```
```

Just the formatting.

- **Advantages**: Maximum speed for queries and updates.
- **Limitations**: RAM usage can become a bottleneck as your dataset grows.

#### 2. Memmap Storage

- **How it works**: Instead of loading all data into memory, memmap storage maps data files directly to a virtual address space on disk. The system's page cache handles data access, making it highly efficient.
- **When to use it**: Perfect for storing large collections that exceed your available RAM while still maintaining near in-memory performance when enough RAM is available.
- **Advantages**: Balances performance and memory usage, allowing you to work with datasets larger than your physical RAM.
- **Limitations**: Slightly slower than pure in-memory storage but significantly more scalable.

To enable memmap vector storage in Qdrant, you can set the **on_disk** parameter to `true` when creating or updating a collection.

```python
client.create_collection(
   collection_name="{collection_name}",
   vectors_config=models.VectorParams(
      …
      on_disk=True
   )
)
```

To do the same for payloads:

> ```python
> client.create_collection(
> collection_name="{collection_name}",
> on_disk_payload= True
> )
> ```

Reviewer: ```suggestion
```python
client.create_collection(
    collection_name="{collection_name}",
    on_disk_payload= True
)
```
```

Formatting

Learn more about [**Sharding in Distributed Deployment**](/documentation/guides/distributed_deployment/)

---

## Query Optimization
![qdrant](/articles_data/vector-search-resource-optimization/query.png)
Improving vector database performance is critical when dealing with large datasets and complex queries. By leveraging techniques like **filtering**, **batch processing**, **reranking**, **rescoring**, and **oversampling**, so you can ensure fast response times and maintain efficiency even at scale.

#### Filtering

Filtering allows you to select only the required fields in your query results. By limiting the output size, you can significantly reduce response time and improve performance.

The filterable vector index is Qdrant's solves pre and post-filtering problems by adding specialized links to the search graph. It aims to maintain the speed advantages of vector search while allowing for precise filtering, addressing the inefficiencies that can occur when applying filters after the vector search.

**Example:**

```python
results = client.search(
    collection_name="my_collection", 
    query_vector=[0.1, 0.2, 0.3], 
    limit=10, 

>     with_payload={"include": ["category"]}

Reviewer 2: this is not a filter

)
```
**Figure 7:** The filterable vector index adds specialized links to the search graph to speed up traversal.

![filterable-vector-index](/articles_data/vector-search-resource-optimization/filterable-vector-index.png)

[**Filterable vector index**](/documentation/concepts/indexing/): This technique builds additional links **(orange)** between leftover data points. The filtered points which stay behind are now traversible once again. Qdrant uses special category-based methods to connect these data points.

---
Read more about [**Filtering Docs**](/documentation/concepts/filtering/) and check out the [**Complete Filtering Guide**](/articles/vector-search-filtering/).

---
#### Batch Processing

Batch processing consolidates multiple operations into a single execution cycle, reducing transaction overhead and enhancing throughput. It’s an effective strategy for both data insertion and query execution.

<img src="/articles_data/vector-search-resource-optimization/batch-processing.png" alt="batch-processing" style="width: 75%;">

> **Batch Insertions**: Instead of inserting vectors individually, group them into larger batches to minimize the number of database transactions and the overhead of frequent writes.

Reviewer 2: Too large batches are not ideal either. Let's call it medium-sized batches.
Also, there are no transactions in Qdrant

            oversampling=2  # Retrieves extra candidates for rescoring
        )
    ),
    limit=4  # Desired number of final results
)
```
---
Learn more about [**Rescoring**](/articles/what-is-vector-quantization/#3-rescoring-with-original-vectors).

---

#### Reranking

Reranking adjusts the order of search results based on additional criteria, ensuring the most relevant results are prioritized. 

This method is about taking the results from different search methods and reordering them based on some additional processing using the content of the documents, not just the scores. This processing may rely on an additional neural model, such as a cross-encoder which would be inefficient enough to be used on the whole dataset. 

<img src="/articles_data/vector-search-resource-optimization/reranking.png" alt="reranking" style="width: 75%;">

These methods are practically applicable only when used on a smaller subset of candidates returned by the faster search methods. Late interaction models, such as ColBERT, are way more efficient in this case, as they can be used to rerank the candidates without the need to access all the documents in the collection.

**Example:**

> ```python
> client.query_points(
>        "collection-name",
>        prefetch=prefetch, # Previous results
>        query=late_vectors, # Colbert converted query
>        using="colbertv2.0",
>        with_payload=True,
>        limit=10,
> )

Reviewer: ``````suggestion
```python
client.query_points(
       "collection-name",
       prefetch=prefetch, # Previous results
       query=late_vectors, # Colbert converted query
       using="colbertv2.0",
       with_payload=True,
       limit=10,
)
```
``````

> ___

Reviewer 2: Snippet is not closed

Learn more about [**Reranking**](/documentation/search-precision/reranking-hybrid-search/#rerank).

---

## Storage: Disk vs RAM
![disk](/articles_data/vector-search-resource-optimization/disk.png)

| Storage  | Description                                                                                                      |
|--------------|------------------------------------------------------------------------------------------------------------------|
| **RAM**      | Crucial for fast access to frequently used data, such as indexed vectors. The amount of RAM required can be estimated based on your dataset size and dimensionality. For example, storing **1 million vectors with 1024 dimensions** would require approximately **5.72 GB of RAM**. |
| **Disk**     | Suitable for less frequently accessed data, such as payloads and non-critical information. Disk-backed storage reduces memory demands but can introduce slight latency. |

#### Which Disk Type?

1. **SSD**: Recommended for optimal performance, particularly for workloads involving random reads and writes. SSDs can significantly enhance query response times when the data is stored on disk.

> 2. **HDD**: While more cost-effective, HDDs are slower and can negatively impact performance, especially for large datasets or applications requiring high-speed access.

Reviewer 2: Almost nobody uses HDD now. The problem of slow disks is not because it is HDD, but because it is a network-attached disk in the cloud, with huge latency

Reviewer 2: Real HDD will be totally unusable with any workload

| **Avoid Re-Sharding**           | Re-sharding can be complex and requires creating a new collection, so it's best to plan ahead to minimize disruptions. |

Learn more about [**Sharding in Distributed Deployment**](/documentation/guides/distributed_deployment/)

---

## Query Optimization
![qdrant](/articles_data/vector-search-resource-optimization/query.png)
Improving vector database performance is critical when dealing with large datasets and complex queries. By leveraging techniques like **filtering**, **batch processing**, **reranking**, **rescoring**, and **oversampling**, so you can ensure fast response times and maintain efficiency even at scale.

#### Filtering

Filtering allows you to select only the required fields in your query results. By limiting the output size, you can significantly reduce response time and improve performance.

The filterable vector index is Qdrant's solves pre and post-filtering problems by adding specialized links to the search graph. It aims to maintain the speed advantages of vector search while allowing for precise filtering, addressing the inefficiencies that can occur when applying filters after the vector search.

**Example:**

```python

> results = client.search(
>     collection_name="my_collection", 
>     query_vector=[0.1, 0.2, 0.3], 
>     limit=10, 
>     with_payload={"include": ["category"]}

Reviewer: ```suggestion
results = client.search(
    collection_name="my_collection",
    query_vector=[0.1, 0.2, 0.3],
    query_filter=models.Filter(must=[
        models.FieldCondition(
            key="category",
            match=models.MatchValue(value="my-category-name"),
        )
    ]),
    limit=10, 
```

)
```
**Figure 7:** The filterable vector index adds specialized links to the search graph to speed up traversal.

![filterable-vector-index](/articles_data/vector-search-resource-optimization/filterable-vector-index.png)

[**Filterable vector index**](/documentation/concepts/indexing/): This technique builds additional links **(orange)** between leftover data points. The filtered points which stay behind are now traversible once again. Qdrant uses special category-based methods to connect these data points.

---
Read more about [**Filtering Docs**](/documentation/concepts/filtering/) and check out the [**Complete Filtering Guide**](/articles/vector-search-filtering/).

---
#### Batch Processing

> Batch processing consolidates multiple operations into a single execution cycle, reducing transaction overhead and enhancing throughput. It’s an effective strategy for both data insertion and query execution.

Reviewer: ```suggestion
Batch processing consolidates multiple operations into a single execution cycle, reducing request overhead and enhancing throughput. It’s an effective strategy for both data insertion and query execution.
```
